visidata/aggregators.py

```diff
@@ -276,10 +307,21 @@ def _fmt_aggr_summary(match, row, trigger_key):
             vd.warning(f'aggregator does not exist: {aggr}')
     return aggrs
 
+@Sheet.api
+def addcol_list_aggr(sheet, col, list_agg):
+    if not sheet.keyCols:
+        vd.fail('ranking requires one or more key columns')
+    c = SettableColumn(name=f'{sheet.name}_{list_agg}', type=int)
+    rows = vd.aggregators[list_agg].aggregate(col, col.sheet.rows)
+    sheet.addColumnAtCursor(c)
+    c.setValues(sheet.rows, *rows)
+
 Sheet.addCommand('+', 'aggregate-col', 'addAggregators([cursorCol], chooseAggregators())', 'Add aggregator to current column')
 Sheet.addCommand('z+', 'memo-aggregate', 'cursorCol.memo_aggregate(chooseAggregators(), selectedRows or rows)', 'memo result of aggregator over values in selected rows for current column')
 ColumnsSheet.addCommand('g+', 'aggregate-cols', 'addAggregators(selectedRows or source[0].nonKeyVisibleCols, chooseAggregators())', 'add aggregators to selected source columns')
+Sheet.addCommand('', 'addcol-rank', 'addcol_list_aggr(cursorCol, "rank")', 'create new column ranking rows by their key columns')
```

Comment:
Is this specific to "rank"?  What happens if we apply a different list aggregator?  Should this take an input?  (the default value could be "rank" to make it the easiest option).

Also what happens with non-list aggregators?  This could be an instant `SUM(curcol) GROUP BY keycols` which I think would be a major hit!

Reply:
> Is this specific to "rank"? What happens if we apply a different list aggregator? Should this take an input? (the default value could be "rank" to make it the easiest option).

I agree, it is more useful to make the command take an input. Right now the only other list aggregator is `list`, and using that would just be equivalent to `=curcol`. It would be excellent for user-defined aggregators though.

> Also what happens with non-list aggregators? This could be an instant `SUM(curcol) GROUP BY keycols` which I think would be a major hit!

This is a great idea! I've pushed a commit for that.

There are two kinds of aggregators, list and non-list. To distinguish them and allow the new column, I've added a field to each aggregator: `listttype`. It specifies what kind of Visidata type the column should have.

Also, by allowing both kinds of aggregators in the same command, what should the behavior be when there are no key columns? For `rank`, we discussed above that perhaps we should use row number, or treat the entire row as a key. But for non-list aggregators, that do `SUM/MIN/MAX/COUNT(curcol) GROUP BY keycols`, neither of those behaviors makes sense. So for now I've stuck with giving an error. I think that helps tutor new users, if they blindly try the command `addcol-aggregator` without knowing the sheet should have key columns.

This PR now creates some conflicts with #2380. I'll focus on getting this PR mergable first, then after it is accepted, I will rework #2380, which I agree is a flawed PR.

Once this PR is in shape to be added, I'll squash it to a smaller set of commits, and fix a tiny conflict with the current develop branch.